### ISSUE TITLE
TASK: add python3-libdnf5 to default install for ansible?

### ISSUE THREAD

**Issue by yselkowitz** [fedora-eln/eln#211]
### What does the ELN SIG need to do?
Following up to the switch to dnf5 (#201), it may be necessary to explicitly add python3 and python3-libdnf5[1] to the ELN installation set, with the exception of -minimal containers (or the like), for the purposes of allowing ansible-driven management in general, as well as package installation in particular, out of the box.  While with dnf-4 the presence of python3 and python3-dnf was a given (and similarly with yum before that), with dnf5 (whose implementation is C++) neither can be assumed.

To that end, we need to determine, and then implement as appropriate:

- Which types of images should contain python3 and python3-libdnf5 by default for the purposes of ansible management?
- The `ansible.default.dnf5` module was added in ansible-core 2.15.  It looks like all supported Fedora versions as well as RHEL/CS 8 and 10 would cover this, but RHEL 9 is still on 2.14.  Do we need to include python3-dnf to support that case (or anything else with an old ansible-core)?
- Are there other Python modules that should be present on such images for commonly-used ansible modules?

[1] https://docs.ansible.com/ansible/latest/collections/ansible/builtin/dnf5_module.html

**Comment by sgallagher** [comment]
@dcantrell Could you help answer some of these questions?

**Comment by yselkowitz** [comment]
Data:
- The ELN base container image does not even include python3
- The ELN generic cloud image includes python3 but not python3-libdnf5 (nor python3-dnf)
- The "default" custom installation option of ELN with Anaconda includes python3 and python3-dnf (because of subscription-manager, see #206) but not python3-libdnf5

**Comment by yselkowitz** [comment]
@dcantrell this is regarding ELN (the future RHEL 11).  Based on the assumption that dnf5 will replace dnf4 in RHEL 11, could you please advise on the above?

**Comment by yselkowitz** [comment]
@dcantrell Right, but based on that, could you please review the questions in the ticket itself and advice accordingly?

**Comment by dcantrell** [comment]
I probably don't know enough about ansible.  Is ansible written in Python?  If so, you will likely need python3-libdnf5 if you want to use dnf5 in ansible.

With dnf5 being rewritten in C++, there are other language binding options.  You can use the C++ library directly and remove the need for Python on the system entirely.  But again, I don't know much about ansible.

**Comment by sgallagher** [comment]
> I probably don't know enough about ansible. Is ansible written in Python? If so, you will likely need python3-libdnf5 if you want to use dnf5 in ansible.

Yes, Ansible is written in Python.
 
> With dnf5 being rewritten in C++, there are other language binding options. You can use the C++ library directly and remove the need for Python on the system entirely. But again, I don't know much about ansible.

Everything Ansible does relies on Python on the target system.

**Comment by dcantrell** [comment]
So then any ELN system with ansible will need python3-libdnf5 and its requirements for dnf5 functionality in ansible.

**Comment by Conan-Kudo** [comment]
This doesn't make sense, because Ansible's raw mode is supposed to be used to bootstrap the target environment. That's why it exists. Ansible is supposed to be able to install its own python dependencies as needed.

**Comment by yselkowitz** [comment]
Well, for container base images that might be true.  But other kinds of images which have python for other reasons may still benefit from having python3-libdnf5 pre-loaded so that a raw mode isn't necessary.

**Comment by sgallagher** [comment]
> Well, for container base images that might be true. But other kinds of images which have python for other reasons may still benefit from having python3-libdnf5 pre-loaded so that a raw mode isn't necessary.

Maybe we can have the `fedora-release-eln` package include `Requires: (python3-libdnf5 if python3)` ? Make it mandatory if python3 exists on the system?

**Comment by Conan-Kudo** [comment]
Wouldn't it make more sense to just have a `Supplements:` instead in `python3-libdnf5`? That way it can be removed if someone doesn't want it.

**Comment by yselkowitz** [comment]
I'm talking about (possibly) adding this to any particular image types where it makes sense, so it could always be removed after the fact.  In general though, this also affects newer Fedora versions and will affect RHEL 11 as well, so I'm not looking to handle it as specific to ELN.

**Comment by yselkowitz** [comment]
See also: https://bugzilla.redhat.com/show_bug.cgi?id=2322751 and https://gitlab.com/testing-farm/infrastructure/-/merge_requests/972
